Iteration 3 - completed

22DPVEmer opened the pull request:
- Added plugin support for dynamic animal type loading that is not part of the main game
- Created Savanna.Common library for shared interfaces and models
- Updated GameField and AnimalFactory to support plugin-based animals, created tiger and zebra plugins for testing
- Refactored test suite to include plugin animal types
- Improved game mechanics with more flexible entity management
- Added entity limit of 200 for the grid
1. New tiger and zebra plugins:
![image](https://github.com/user-attachments/assets/27b7f15c-fd02-4769-9394-77b95a3fd7b7)
2. Entity limiter:
![image](https://github.com/user-attachments/assets/8f303497-025e-42bb-be97-660a82150952)
3. Updated passing test cases:
![image](https://github.com/user-attachments/assets/808c0cbd-5d22-4c75-ae9f-e32adf38c1fc)

22DPVEmer committed:
Implement plugin system and enhance game architecture

- Added plugin support for dynamic animal type loading
- Created Savanna.Common library for shared interfaces and models
- Updated GameField and AnimalFactory to support plugin-based animals, created tiger and zebra plugins for testing
- Refactored test suite to include plugin animal types
- Improved game mechanics with more flexible entity management
- Seperated commonly used interfaces in a seperate project
- Added entity limit 200 for the grid

22DPVEmer committed:
Improve game entity management and error handling

- Refactored GameField to filter only alive entities in queries
- Added cleanup of dead animals before adding new ones
- Implemented more robust entity limit handling in GetActiveEntities
- Removed debug console output from Lion class
- Updated error handling in GameRunner to silently ignore invalid animal types
- Added a centralized unknown animal type message in TestConstants

sreeharihm commented on Savanna.ConsoleApp/Game/GameRunner.cs:
```diff
@@ -68,33 +70,36 @@ private void RunGameLoop()
-            var key = Console.ReadKey(true).Key;
-            switch (key)
+            var key = Console.ReadKey(true).KeyChar;
+            if (key == 'q' || key == 'Q')
```
Move these to constant

22DPVEmer replied:
I moved it.

sreeharihm commented on Savanna.Plugins.Tiger/Tiger.cs:
```diff
@@ -0,0 +1,132 @@
+            return field.GetEntitiesInRange(Position, VisionRange)
+                .Where(entity =>
+                    entity.Symbol != LION_SYMBOL &&  // Don't attack lions
+                    (entity.Symbol == 'Z' || entity.Symbol == 'A') &&  // Attack Zebras or Antelopes
```
Symbols should be in a constant class

22DPVEmer replied:
I moved them to a constants class.

sreeharihm commented on Savanna.Plugins.Tiger/TigerConfiguration.cs:
```diff
@@ -0,0 +1,15 @@
+    /// </summary>
+    public class TigerConfiguration : IAnimalConfiguration
+    {
+        public int Speed => 2;  // Slower but stronger
```
Same here, move to constants

22DPVEmer replied:
Moved the value to a constant class.

sreeharihm commented on Savanna.Plugins.Tiger/TigerPlugin.cs:
```diff
@@ -0,0 +1,27 @@
+        private readonly TigerConfiguration _configuration = new();
+
+        public string AnimalName => "Tiger";
+        public char Symbol => _configuration.Symbol;
```
Same here, move to constant

22DPVEmer replied:
Moved it to a constant class.

sreeharihm commented on Savanna.Plugins.Zebra/Zebra.cs:
```diff
@@ -0,0 +1,92 @@
+
+            // Look for predators and try to escape if they're nearby
+            var nearestPredator = field.GetEntitiesInRange(Position, VisionRange)
+                .Where(e => (e.Symbol == 'L' || e.Symbol == 'T') && e.IsAlive)
```
Move these symbols to a constant or an enum

sreeharihm commented on Savanna.Plugins.Zebra/ZebraPlugin.cs:
Move to constant class

22DPVEmer replied:
I moved the values to a constant class.

sreeharihm reviewed (commented):
Please move all the hardcoded values to a constant class

22DPVEmer committed:
Centralize constants and improve code consistency across plugins

- Added GameConstants for shared plugin and user interface constants
- Extracted hardcoded values to constants in Tiger, Zebra, and base plugin classes
- Standardized configuration, symbol, and plugin metadata constants
- Removed magic numbers and strings in favor of centralized constant definitions
- Improved code readability and maintainability by using consistent constant references

sreeharihm commented on Savanna.Plugins.Zebra/Constants/ZebraConstants.cs:
As an improvement, you can remove configration constant and use a json file for these configuration. Once application up, load those configuration from the json file and do the rest of the operation

22DPVEmer replied:
I created the json files for both plugins and made it load on game start

22DPVEmer committed:
Migrate Tiger and Zebra plugins to JSON configuration

- Replace hardcoded constants with dynamic JSON configuration
- Add config.json file support for Tiger and Zebra plugins
- Implement configuration loading with fallback and error handling
- Update plugin classes to use configuration from JSON
- Remove deprecated constant classes
- Add configuration model classes for type-safe configuration

sreeharihm commented on Savanna.Plugins.Zebra/config.json:
I would prefer for a single JSON file, that should be shared across the project

22DPVEmer replied:
Refactored it to now handle the single JSON from the Common project across both Zebra and Tiger plugins.

22DPVEmer committed:
Refactor plugin configuration and remove local config files

- Move configuration models from plugin-specific Models folders to Savanna.Common
- Remove local config.json files from Tiger and Zebra plugins
- Centralize configuration loading in Savanna.Common
- Simplify plugin configuration initialization
